journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java

```diff
@@ -493,57 +496,110 @@ private FileChannel openChannel(final File file) {
-        final JournalSegmentFile segmentFile = new JournalSegmentFile(file);
-        final ByteBuffer buffer = ByteBuffer.allocate(JournalSegmentDescriptor.getEncodingLength());
-        try (final FileChannel channel = openChannel(file)) {
-          channel.read(buffer);
-          buffer.flip();
-        } catch (final IOException e) {
-          throw new JournalException(e);
+        if (i > 0 && segments.get(i - 1).lastIndex() + 1 != segment.index()) {
+          throw new CorruptedLogException(
+              String.format(
+                  "Log segment %s is not aligned with previous segment %s.",
+                  segments.get(i - 1), segment));
         }
 
-        final JournalSegmentDescriptor descriptor = new JournalSegmentDescriptor(buffer);
-
-        // Load the segment.
-        final JournalSegment segment = loadSegment(descriptor.id());
+        log.debug("Found segment: {} ({})", descriptor.id(), file.getName());
+        segments.add(segment);
+      } catch (final CorruptedLogException e) {
+        if (handleSegmentCorruption(files, segments, i)) {
+          return segments;
+        }
 
-        // Add the segment to the segments list.
-        log.debug(
-            "Found segment: {} ({})", segment.descriptor().id(), segmentFile.file().getName());
-        segments.put(segment.index(), segment);
+        throw e;
       }
     }
 
-    // Verify that all the segments in the log align with one another.
-    JournalSegment previousSegment = null;
-    boolean corrupted = false;
-    final Iterator<Entry<Long, JournalSegment>> iterator = segments.entrySet().iterator();
-    while (iterator.hasNext()) {
-      final JournalSegment segment = iterator.next().getValue();
-      if (previousSegment != null && previousSegment.lastIndex() != segment.index() - 1) {
-        log.warn(
-            "Journal is inconsistent. {} is not aligned with prior segment {}",
-            segment.file().file(),
-            previousSegment.file().file());
-        corrupted = true;
+    return segments;
+  }
+
+  /** Returns true if segments after corrupted segment were deleted; false, otherwise */
+  private boolean handleSegmentCorruption(
+      final List<File> files, final List<JournalSegment> segments, final int failedIndex) {
+    long lastSegmentIndex = 0;
+
+    if (!segments.isEmpty()) {
+      final JournalSegment previousSegment = segments.get(segments.size() - 1);
+      lastSegmentIndex = previousSegment.lastIndex();
+    }
+
+    if (lastWrittenIndex > lastSegmentIndex) {
+      return false;
+    }
+
+    log.debug(
+        "Found corrupted segment after last ack'ed index {}. Deleting it and subsequent segments",
+        lastWrittenIndex);
+
+    for (int i = failedIndex; i < files.size(); i++) {
+      final File file = files.get(i);
+      try {
+        Files.delete(file.toPath());
+      } catch (IOException e) {
+        log.warn("Failed to delete log segment {}", file.getName(), e);
```

Comment:
Would it create problem later if we are not able to delete the segment properly? What would happen when we have to create a new segment with the same id?

Reply:
AFAIK, mmap will open it and the journal will overwrite its contents. We can also just fail hard here to be safe. In the next restart there will be another attempt to delete it since it's the corrupted file or it's one after it and there will be a gap in the indexes

Reply:
It might be safe to fail hard, I think. Consider the following scenario:

Segment {id = 1, firstIndex = 1, lastIndex = 1}
Segment {id = 2, firstIndex = 2, lastIndex = 2} - corrupted
Segment {id = 3, firstIndex = 3, lastIndex = 3}

1. Deleted Segment-2
2. Failed to delete Segment-3. 
3. append record at index 2
4. restart broker
5. Load segments - segment 3 is loaded successfuly, no gap in indexes, but the record at index 3 is incorrect.